feat: add support for multiple API keys in addition to the user key

monomarh opened the pull request:
Closes #359 — Support for multiple API tokens. 
Also includes enhancement requested by @wfjsw — Read-only token support for shared data access.

muety commented on config/config.go:
```diff
@@ -544,6 +543,7 @@ func Load(configFlag string, version string) *Config {
 	config.InstanceId = uuid.Must(uuid.NewV4()).String()
 	config.App.Colors = readColors()
 	config.Db.Dialect = resolveDbDialect(config.Db.Type)
+	slog.Info("loaded configuration", "environment", config.Env, "version", config.Version, "db_dialect", config.Db.Type)
```
Why put this log line just exactly here, in the middle of config loading? Why not at the end?

muety commented on middlewares/authenticate.go:
```diff
@@ -65,6 +67,11 @@ func (m *AuthenticateMiddleware) WithRedirectErrorMessage(message string) *Authe
 }
 
+func (m *AuthenticateMiddleware) WithOnlyRWApiKeys(onlyRW bool) *AuthenticateMiddleware {
+	m.onlyRWApiKeys = onlyRW
```
This attribute is not actually being checked somewhere, is it?

muety commented on services/user.go:
```diff
@@ -101,17 +104,48 @@ func (srv *UserService) GetUserByKey(key string) (*models.User, error) {
 		return nil, errors.New("key must not be empty")
 	}
 
+	// TODO: does it make sense to use cache by api key if it's not being used?
```
Fair point. Please delete these lines.

muety commented on Dockerfile:
```diff
@@ -1,26 +1,33 @@
 
-RUN wget "https://raw.githubusercontent.com/vishnubob/wait-for-it/master/wait-for-it.sh" -O wait-for-it.sh && \
-    chmod +x wait-for-it.sh
+RUN wget "https://raw.githubusercontent.com/vishnubob/wait-for-it/master/wait-for-it.sh" -O /bin/wait-for-it.sh && \
```
Please explain.

muety commented on models/api_key.go:
```diff
@@ -0,0 +1,14 @@
+package models
+
+type ApiKey struct {
```
We now have API keys stored in two different places. There is the "main" API key as part of the user model and additional API keys in form of this many-to-many relation. I think that's somewhat confusing both for a user and for us as developers. What do you think?

muety reviewed (changes requested):
Thanks a lot for this contribution! Looks like it's still very much in progress, though?

Apart from that, I'd really appreciate contributors to first have a look at our [guidelines](https://github.com/muety/wakapi/wiki/Contributing) before pushing a PR. It would be great if people first double-checked on an issue for whether they may go about implementing it before jumping straight in.

muety commented:
Any updates on this?

monomarh commented:
> Any updates on this?

Hey man, it's been a tough week :) I fixed all your comments and removed all the junk code I added to enable hot reload and dlv. Thank you for having the patience to look through it all. I didn't notice that VS code added it all to the index itself. By the way, I haven't programmed on GO before, so I would really appreciate any comments you may have!

muety commented:
Thanks for implementing all those changes! And sorry that it took me another week to get back to this now. I think it's a great contribution, especially considering that it's your first time contributing to a Go project :+1:.

The only one last thing I'd like to ask you for is to add some unit tests for this new functionality, i.e. primarily extend the authentication middleware's set of test cases. Add positive and negative test cases for authenticating with a "non-main" API key and please also add a test that validates the rw- vs read-only logic. 

After that's done, the PR is ready to be merged.

monomarh commented on middlewares/authenticate.go:
```diff
@@ -65,6 +67,11 @@ func (m *AuthenticateMiddleware) WithRedirectErrorMessage(message string) *Authe
 	return m
 }
 
+func (m *AuthenticateMiddleware) WithFullAccessOnly(readOnly bool) *AuthenticateMiddleware {
```
thanks, updated
